nntrainer/src/optimizer.cpp

```diff
@@ -65,64 +65,71 @@ int Optimizer::initialize(TensorDim d, bool set_tensor) {
     return ML_ERROR_INVALID_PARAMETER;
   }
   if (type == OptType::adam && set_tensor) {
-    wm = Tensor(d.channel(), d.height(), d.width());
-    wv = Tensor(d.channel(), d.height(), d.width());
+    Tensor wm = Tensor(d.channel(), d.height(), d.width());
+    Tensor wv = Tensor(d.channel(), d.height(), d.width());
     wm.setZero();
     wv.setZero();
-    bm = Tensor(1, 1, d.width());
-    bv = Tensor(1, 1, d.width());
+    weight_mv.push_back(std::pair<Tensor, Tensor>(wm, wv));
+
+    Tensor bm = Tensor(1, 1, d.width());
+    Tensor bv = Tensor(1, 1, d.width());
     bm.setZero();
     bv.setZero();
+    weight_mv.push_back(std::pair<Tensor, Tensor>(bm, bv));
   }
   return status;
 }
 
-void Optimizer::calculate(const Tensor &djdw, const Tensor &djdb,
-    Tensor &weight, Tensor &bias, int iteration) {
-  Tensor djdwAvg, djdbAvg;
+void Optimizer::calculate_gradients(
+    std::vector<std::pair<Tensor, Tensor>> val_grad, int iteration) {
   float ll = popt.learning_rate;
   if (popt.decay_steps != -1) {
     ll = ll * pow(popt.decay_rate, (iteration / popt.decay_steps));
   }
 
-  djdwAvg = djdw.average();
-  djdbAvg = djdb.average();
-
-  switch (type) {
-  case OptType::sgd:
-    weight.add_i(djdwAvg, -ll);
-    bias.add_i(djdbAvg, -ll);
-    break;
-  case OptType::adam: {
-    std::function<float(float)> sqrtEps = [&](float f) {
-      return sqrtFloat(f) + this->popt.epsilon;
-    };
+  if (type == OptType::adam) {
     std::function<float(float)> biasCorrection = [&](float f) {
       return 1 - pow(f, iteration + 1);
     };
 
     ll *= sqrt(biasCorrection(popt.beta2)) / biasCorrection(popt.beta1);
+  }
 
-    wm.multiply_i(popt.beta1);
-    wm.add_i(djdwAvg, 1 - popt.beta1);
+  int idx = 0;
+  for (auto iter = val_grad.begin(); iter != val_grad.end(); iter++) {
+    Tensor x = (*iter).first;
+    Tensor x_grad = (*iter).second;
 
-    wv.multiply_i(popt.beta2);
-    wv.add_i(djdwAvg.multiply(djdwAvg), 1 - popt.beta2);
+    x_grad = x_grad.average();
 
-    weight.add_i(wm.divide(wv.apply(sqrtEps)), -ll);
+    switch (type) {
+    case OptType::sgd:
+      x.add_i(x_grad, -ll);
+      break;
+    case OptType::adam: {
+      std::function<float(float)> sqrtEps = [&](float f) {
+        return sqrtFloat(f) + this->popt.epsilon;
+      };
 
-    bm.multiply_i(popt.beta1);
-    bm.add_i(djdbAvg, 1 - popt.beta1);
+      Tensor wm = weight_mv[idx].first;
+      Tensor wv = weight_mv[idx].second;
 
-    bv.multiply_i(popt.beta2);
-    bv.add_i(djdbAvg.multiply(djdbAvg), 1 - popt.beta2);
+      wm.multiply_i(popt.beta1);
+      wm.add_i(x_grad, 1 - popt.beta1);
 
-    bias.add_i(bm.divide(bv.apply(sqrtEps)), -ll);
-    break;
-  }
-  case OptType::unknown:
-  default:
-    break;
+      wv.multiply_i(popt.beta2);
+      wv.add_i(x_grad.multiply(x_grad), 1 - popt.beta2);
+
+      x.add_i(wm.divide(wv.apply(sqrtEps)), -ll);
+      break;
```

Comment:
bias update?

Reply:
This operates over the list of weights. So, the caller function will add all the weights and bias as a list
https://github.com/nnstreamer/nntrainer/blob/a112a6064633e7495fd5cdd882a59a1483d213d7/nntrainer/src/fc_layer.cpp#L175-L176

Reply:
I see. I missed it.